transport/handler_server.go

```diff
@@ -319,7 +320,16 @@ func (ht *serverHandlerTransport) HandleStreams(startStream func(*Stream)) {
 				s.buf.put(&recvMsg{data: buf[:n]})
 			}
 			if err != nil {
-				s.buf.put(&recvMsg{err: err})
+				if err, ok := err.(http2.StreamError); ok {
+					code := http2ErrConvTab[err.Code]
+					if code == codes.OK {
+						s.buf.put(&recvMsg{err: io.EOF})
+					} else {
+						s.buf.put(&recvMsg{err: StreamErrorf(code, "%v", err)})
+					}
+				} else {
+					s.buf.put(&recvMsg{err: io.EOF})
```

Comment:
@bradfitz 

I observed some strange behavior here from http2 package. In some end2end tests (e.g., TestPingPong_4), this branch is entered when err is non-nil with "stream error: stream ID 0; NO_ERROR" and its type is not http2.StreamError. And the expected status in the test is OK (i.e., io.EOF). Here I forced it return a success (i.e., io.EOF). But this sounds strange to me and I am not sure it is the right thing to do. Can you take a look?